### DISCUSSION TITLE
Let's ship it?

### DISCUSSION THREAD

**Discussion by mk6i** [supermetrics/php-ext-jsonpath#120] [General]
@crocodele What are your thoughts for next steps? I was thinking the next two milestones would be for Supermetrics to dogfood `php-ext-jsonpath` and then submit to PECL once we're confident that it's production-grade.

**Comment by crocodele** [comment]
@mkaminski1988, yes, I will apply for the PECL account to get the process started. :slightly_smiling_face:

**Reply by crocodele** [reply]
I submitted the request to the PECL list, [here's the initial feedback](https://news-web.php.net/php.pecl.dev/17406).

I suppose we should wrap everything in a namespace, and if I remember correctly, it was an item in the original issue list too. :slightly_smiling_face: The [namespacing rules](https://wiki.php.net/rfc/namespaces_in_bundled_extensions) that Benjamin refers to state that `the top-level namespace should match the extension name (apart from casing)`, so I guess that would be `JsonPath` then.

As for the array vs. string comment, I think it's a valid point. Perhaps we could provide two methods: `findFromArray(array $data, string $selector)` and `findFromJson(string $jsonString, string $selector)` instead of a single `find(array $data, string $selector)`? At least for now, `findFromJson()` could simply do `json_decode()` + `findFromArray()`.

@mkaminski1988, what do you think?

**Reply by mk6i** [reply]
Thanks @crocodele, you worded the request really well.

I'm in favor of implementing both suggestions.

What if we treated the `$data` param in `find()` as a mixed type that accepts both strings and arrays, and the implementation calls `json_decode()` when `$data` is a string? (I'm out of the loop on PHP best practices, maybe that's no longer idiomatic?)

**Reply by Danack** [reply]
> as a mixed type that accepts both strings and arrays, ....I'm out of the loop on PHP best practices

@mkaminski1988 , as fond as I am of the mixed type, using a union type of `string|array` would be more precise for what that parameter accepts.

> Perhaps we could provide two methods findFromArray ... and ... findFromJson

That sounds sensible. Although, you could keep the `find(array|string $data, string $selector)` if you wanted to, I'm finding it hard to imagine in the code I write, where I would be unsure if I had either an array, or a JSON encoded string.

**Reply by crocodele** [reply]
There are a couple of reasons why I would prefer specific methods rather than one-method-to-rule-them-all:

1. In general, I like strict and non-ambiguous types, and explicit methods for specific purposes.
2. If we decide to add support for more sources, we're not restricted by the `find(array|string|SomeJsonDataSource $data, string $selector)` method signature, but can for example add more parameters if needed. Not sure what sources that would be, but you never know. :slightly_smiling_face: 
3. When we decode the JSON string, I think we should throw a `JsonException` if decoding fails. If we have a single `find()` method, the docblock for that would be:
```
/**
 * @param array|string $data
 * @param string       $selector
 *
 * @return array|bool
 *
 * @throws \JsonException
 * @throws \RuntimeException (at least for now, maybe a custom exception later)
 */
public function find(array|string $data, string $selector): array|bool;
```

But the `JsonException` doesn't apply to the array case, only to the string case, and I wouldn't want my IDE warning me about uncaught `\JsonException` if I pass in an array.

This is just me though; happy to hear other thoughts. :slightly_smiling_face:

**Reply by mk6i** [reply]
Thanks for your input @crocodele @Danack.

I think Viktor's points make sense from a user perspective. As a maintainer, I have a mild preference for the status quo for a couple reasons:

- Calling `json_decode()` from within the extension increases the extension's scope of responsibility. If this was a PHP library, that wouldn't be an issue, because the code would depend on a stable API. For the extension implementation, I believe that we'll need to call the internal [`php_json_encode_ex()`](https://github.com/php/php-src/blob/master/ext/json/json.c#L144) in order to decode the JSON, which increases the surface area of things that could break in future PHP versions (I don't believe we can call the public `PHP_FUNCTION(json_encode)` directly, but I could be wrong about that.) In addition, we have to take care to properly proxy and document errors emitted from `json_decode()` through the extension.
- To me it seems a bit of an anti-pattern for a PHP extension to depend on another PHP extension. In practical terms, users of PHP 7.4 must not exclude the `json` extension from the runtime. This won't be an issue for PHP 8, however, as [the extension is required](https://www.php.net/manual/en/json.installation.php#json.installation). Again, another thing to document and maintain. (Edit: The extension already depends on the `pcre` and `spl` extensions, so this point is probably moot)

I say that you should make the final call @crocodele, as you're a stakeholder that will actually use this. :) I support either approach.

**Reply by crocodele** [reply]
Let's keep things as-is for now and come back to this later if we decide to add support for more sources (filepaths, file resources, streams, ...).

**Reply by mk6i** [reply]
Any news on the submission process?

**Reply by crocodele** [reply]
It's a while since I applied for the PECL account, but I haven't heard back yet. I'll wait a few more days and then I'll ask in the newsgroup.

On a completely different topic, could you @mkaminski1988 please send me a ping to viktor.djupsjobacka@supermetrics.com so I get your e-mail address?

**Reply by mk6i** [reply]
Thanks for the update, maybe everyone's on vacation!

Sent you an email.

**Reply by Danack** [reply]
> It's a while since I applied for the PECL 

@crocodele fyi, I don't think the application was processed. You may have failed the captcha which is deliberately obtuse, for quite bad reasons.

I say that the application wasn't processed, as I can't see an account request (which looks like https://news-web.php.net/php.pecl.dev/17353) in the mailing list: https://news-web.php.net/group.php?group=php.pecl.dev

Feel free to put me as the sponsor.

**Reply by mk6i** [reply]
Thanks for the info @Danack 🌮

**Reply by crocodele** [reply]
@Danack, the application must have ended up in some weird limbo state. When I try to submit the form now with the same username and e-mail address, I get `Sorry, that username is already taken` and `Sorry, that email is already registered in the database`. Can you find and enable this account, or should I try again with another username and e-mail address?

Username: `viktor`
E-mail address: viktor.djupsjobacka@supermetrics.com

![Screenshot from 2021-08-06 21-21-27](https://user-images.githubusercontent.com/1522694/128555541-a5faa5d2-c9e8-4ef6-a1fc-24965a4fb4e1.png)

**Comment by mk6i** [comment]
Regarding [Dan Ackroyd's comment](https://news-web.php.net/php.pecl.dev/17408):

> Hi Viktor,
> 
> For entertainment, I checked out the extension and ran the tests with
> valgrind i.e. with "php run-tests.php -m -j4". There seem to be a few
> invalid reads for a couple of the tests; output is below.

I haven't looked at each test case that causes the errors he reported. For the ones I _did_ look at, commenting out calls to `php_pcre_match_impl()` in the interpreter "fixed" the error (even if the test failed).

I think there might be a bug in `php_pcre_match_impl()` itself. For example, I can reproduce half of the example error provided by Dan with the following PHP script, which does not execute the JSONPath extension code.

```php
<?php

preg_match('/(a)(b)*(c)/', 'xxxxxx', $matches);
var_dump($matches);
```

```sh
USE_ZEND_ALLOC=0 valgrind --leak-check=full php test.php
```

Which outputs the following for me on php 8.0.8/gcc 7.5.0:

```
==14692== Conditional jump or move depends on uninitialised value(s)
==14692==    at 0x403AAAD: ???
==14692==    by 0x932DC67: ???
==14692==    by 0x932DC67: ???
==14692==    by 0x932DC6D: ???
==14692==    by 0x91301EF: ???
==14692==    by 0x932DC67: ???
```

cc @danack

**Reply by Danack** [reply]
Bug reported: https://bugs.php.net/bug.php?id=81240

As enabling debug 'makes this go away', there's a reasonable chance that it's a bug in PHP rather than the underlying pcre library.

**Reply by crocodele** [reply]
Thank you both! :heart:

**Comment by Danack** [comment]
@mkaminski1988 in case you didn't see the update on what I thought was a bug - it turns out when you enable debug in PHP, that enables compiling with valgrind support, which stops valgrind from issuing false reports. 

So as there's no visible problem, and the valgrind report goes away in the debug build, then it's more likely to be a false valgrind report, rather than an actual issue. Probably.

**Comment by cmb69** [comment]
My apologies regarding not having approved the PECL account request of @crocodele earlier. Apparently, the notification mail had not been sent. If you have any questions regarding the PECL account, feel free to ask on the mailing list, or just ping me on GH. Happy PECLing! :)

**Reply by crocodele** [reply]
Thank you @cmb69 and @Danack! Managed to log in now, register the package, and publish the first release. :tada:

**Comment by mk6i** [comment]
Wow, this is huge! Nice work, Viktor!

Do I need to register for a PECL account to be listed as a maintainer as well?

**Reply by crocodele** [reply]
@mkaminski1988, yup, you need a PECL account too.

**Reply by cmb69** [reply]
Then goto https://pecl.php.net/account-request.php. :)

**Reply by mk6i** [reply]
Submitted :)

**Reply by cmb69** [reply]
Approved. :)
